akka-stream/src/main/scala/akka/stream/impl/Timers.scala

```diff
@@ -276,11 +276,15 @@ import akka.stream.stage._
           val now = System.nanoTime()
           // Timer is reliably cancelled if a regular element arrives first. Scheduler rather schedules too late
           // than too early so the deadline must have passed at this time.
-          assert(
-            now - nextDeadline >= 0,
-            s"Timer should have triggered only after deadline but now is $now and deadline was $nextDeadline diff ${now - nextDeadline}.")
-          push(out, inject())
-          nextDeadline = now + timeout.toNanos
+          val diff = now - nextDeadline
+          if (diff < 0) {
+            log.error(
+              s"Timer should have triggered only after deadline but now is $now and deadline was $nextDeadline diff $diff.")
+            scheduleOnce(GraphStageLogicTimer, FiniteDuration(-diff, TimeUnit.NANOSECONDS))
```

Comment:
logging an error message and rescheduling when triggered *early*.

Reply:
Make it a warning instead, it is not really critical.